nucypher/acumen/pruning.py

```diff
@@ -0,0 +1,63 @@
+"""
+ This file is part of nucypher.
+
+ nucypher is free software: you can redistribute it and/or modify
+ it under the terms of the GNU Affero General Public License as published by
+ the Free Software Foundation, either version 3 of the License, or
+ (at your option) any later version.
+
+ nucypher is distributed in the hope that it will be useful,
+ but WITHOUT ANY WARRANTY; without even the implied warranty of
+ MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ GNU Affero General Public License for more details.
+
+ You should have received a copy of the GNU Affero General Public License
+ along with nucypher.  If not, see <https://www.gnu.org/licenses/>.
+"""
+
+from collections import Counter
+
+import maya
+from typing import Optional, Callable
+
+
+def accept_node(node: "Teacher", reset: bool = False) -> bool:
+    """
+    Return True to keep a node or False to Erase
+    NOTE: This function *must* return a boolean to operate properly.
+    """
+    return True
+
+
+def reject_node(node: "Teacher", reset: bool = False) -> bool:
+    """Indicate a node can be deleted now."""
+    return False
+
+
+def construct_node_stalecheck(max_seconds: int, start_time: Optional[maya.MayaDT] = None) -> Callable:
+    start_time = start_time or maya.now()
+
+    def check(node: "Teacher", reset: bool = False) -> bool:
```

Comment:
Not sure how I feel about that. As much as I like FP, would a `Strategy` ABC be more understandable in code?

Reply:
Initially @vepkenez and I were compelled to keep these strategies completely stateless, but I'm not sure how that's *truly possible* if we're going to track an a per-node basis.  There might be some way to rework the rules of invalidation with statelessness in mind... :thinking:

Reply:
Hm, well, for example you could do the following:
- Have a single dict (`node -> label`) (or `checksum_address -> label`). A label here is an object that incorporates a strategy.
- In `prune_bucket()`, you iterate over nodes and `__call__` the label of each. It either returns another label (or the same one), or `None`, the latter meaning that the node is to be removed.

This way you can have a `MaxAttempts` label that will return a `MaxAttempts(label.count - 1)` on each call, or `None` if `count` is 0. Similarly for `Stalecheck`.

Reply:
Of course, you'll lose the ability to quickly iterate over nodes with a chosen label. You'll have to maintain supporting structures if that's desirable.

Reply:
Reworked as a callable ABC, although much more simply than described above.  There may need to be more development iterations but I've enough ground work to continue on.